spacy_experimental/span_boundary_detection/sbd_component.py

```diff
@@ -0,0 +1,301 @@
+            scorer_start.tp += 1
+        elif start_prediction == 1 and start_reference == 0:
+            scorer_start.fp += 1
+        elif start_prediction == 0 and start_reference == 1:
+            scorer_start.fn += 1
+
+        # End
+        if end_prediction == 1 and end_reference == 1:
+            scorer_end.tp += 1
+        elif end_prediction == 1 and end_reference == 0:
+            scorer_end.fp += 1
+        elif end_prediction == 0 and end_reference == 1:
+            scorer_end.fn += 1
+
+    # Assemble final result
+    final_scores: Dict[str, Any] = {
+        f"sbd_start_f": scorer_start.fscore,
+        f"sbd_start_p": scorer_start.precision,
+        f"sbd_start_r": scorer_start.recall,
+        f"sbd_end_f": scorer_end.fscore,
+        f"sbd_end_p": scorer_end.precision,
+        f"sbd_end_r": scorer_end.recall,
+    }
+
+    return final_scores
+
+
+def get_reference(docs) -> Floats2d:
+    """Create a reference list of token probabilities for calculating loss and metrics"""
+    reference_results = []
+    for doc in docs:
+        start_indices = []
+        end_indices = []
+
+        for spankey in doc.spans:
+            for span in doc.spans[spankey]:
+                start_indices.append(span.start)
+                end_indices.append(span.end - 1)
+
+        for token in doc:
+            is_start = 0
+            is_end = 0
+            if token.i in start_indices:
+                is_start = 1
+            if token.i in end_indices:
+                is_end = 1
+            reference_results.append((is_start, is_end))
+
+    return reference_results
+
+
+def get_predictions(docs) -> Floats2d:
+    """Create a prediction list of token start/end probabilities for evaluation"""
+    prediction_results = []
+    for doc in docs:
+        for token in doc:
+            prediction_results.append([token._.span_start, token._.span_end])
+    return prediction_results
+
+
+class SpanBoundaryDetection(TrainablePipe):
+    """Pipeline that learns start and end tokens of spans"""
+
+    def __init__(
+        self,
+        nlp: Language,
+        model: Model[List[Doc], Floats2d],
+        name: str = "sbd",
+        *,
+        threshold: float = 0.5,
+        scorer: Optional[Callable] = sbd_score,
+    ) -> None:
+        """Initialize the span boundary detector.
+        model (thinc.api.Model): The Thinc Model powering the pipeline component.
+        name (str): The component instance name, used to add entries to the
+            losses during training.
+        threshold (float): Minimum probability to consider a prediction
+            positive.
+        scorer (Optional[Callable]): The scoring method.
+        """
+        self.vocab = nlp
+        self.cfg = {
+            "threshold": threshold,
+        }
+        self.model = model
+        self.name = name
+        self.scorer = scorer
+        Token.set_extension("span_start", default=0, force=True)
```

Comment:
Currently for the `iob` tags the `Token` has attributes `Token.ent_iob` and `Token.ent_iob_` https://spacy.io/api/token#attributes. It seems to me that each `Token` could be either `B`, `E` or nothing so maybe `Token.span_boe`? Not sure

Reply:
So you mean adding a new attribute to the Token? `Token.span_boe` 😄

Reply:
yeah something like that, rather than having a `span_start` and a `span_end` it would be more like the way the `Token.ent_iob` works.

Reply:
There are a lot of drawbacks to representing span annotation as token attributes I'd prefer not to add any more of it, and also drawbacks to how integers are used internally for `ent_iob` in particular.

What are the exact requirements here?

Reply:
I'll rebuild the whole architecture so that the SpanFinder assigns a SpanGroup with all predicted spans and the SpanFinderSuggester uses the SpanGroup to create the Ragged object for the spancat prediction. This way we don't need another token attribute 😄